### DISCUSSION TITLE
Visualizing individual light rays

### DISCUSSION THREAD

**Discussion by h-OUS-e** [mitsuba-renderer/mitsuba3#523] [Q&A] [answered]
Not really an issue but just a question, what is the best way to visualize individual light rays coming out of the emitter, being refracted, and hitting a camera?

thanks!

**Comment by njroussel** [comment, marked as answer]
You can write a custom integrator which only samples lights paths according to your pre-defined behaviour. That would allow you to get a rendering where you only see the contribution in the final image of those light paths.

If you want to get individual light paths (handful), then you can simulate them "by-hand" by scripting the integrator for those few rays which would allow you to effectively save the bounces to some data structure which you can later use in your visualization.

**Reply by h-OUS-e** [reply]
Thank you! And following up with a couple of questions.

1. Is there any reference on how to write your own integrator plugin? Or an example of a custom integrator written in python instead of C+ to be able to write one from there?

2. Is the particle tracer (ptracer) the only integrator where light rays are emitted from the source rather than the camera?

3. How would I script the individual light rays in the integrator? don't these rays come out of the emitter?

4. What would be the easiest way to calculate the number of rays coming out of the emitter, and the number of those rays that have made it to some surface.

**Reply by njroussel** [reply]
1. We have a few. Here's a simple one: https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/ad/integrators/direct_reparam.py. If you're not interested in differentiation you can inherit of `mi.SamplingIntegrator`. You only need to implement the `sample()` method.

2. Yes.

3. In [this tutorial](https://mitsuba.readthedocs.io/en/stable/src/rendering/scripting_renderer.html) we essentially implement an Integrator and some of the boilerplate around it without having to deal with the whole `mi.SamplingIntegrator` interface. This is what I meant, by scripting something specifically for the behaviour you want.

4. Keep a counter per shape. You can get the shape information inside of `si.shape` where `si` is a `mi.SurfaceInteraction3f` which you'll typically get after a ray intersection.

**Reply by h-OUS-e** [reply]
That is super helpful thank you!
